File: android/app/src/main/java/com/ugid/scanner/ApiService.kt
package com.ugid.scanner

import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import retrofit2.http.Body
import retrofit2.http.POST

data class ParseRequest(val payload: String)

data class FingerprintModel(
    val finger_index: Int?,
    val minutiae_count: Int?,
    val minutiae_bytes: Int?,
    val sealed_block_bytes: Int?
)

data class ParseResponse(
    val surname: String,
    val given_name: String,
    val other_name: String,
    val full_name: String,
    val date_of_birth: String,
    val issue_date: String,
    val expiry_date: String,
    val nin: String,
    val sex: String,
    val card_number: String,
    val age: Int,
    val is_expired: Boolean,
    val fingerprint: FingerprintModel,
    val warnings: List<String>
)

interface ApiService {
    @POST("parse")
    suspend fun parseBarcode(@Body request: ParseRequest): Response<ParseResponse>
}

object RetrofitClient {
    // EMULATOR: use 10.0.2.2:8000
    // PHYSICAL DEVICE: use your computer's WiFi IP, e.g., "http://192.168.1.45:8000/"
    private const val BASE_URL = "http://10.0.2.2:8000/"

    val apiService: ApiService by lazy {
        Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())
            .build()
            .create(ApiService::class.java)
    }
}

File: ios/UgandaIDScanner/APIService.swift
import Foundation

struct ParseRequest: Codable {
    let payload: String
}

struct ParseResponse: Codable, Identifiable {
    let id = UUID()
    let surname: String
    let given_name: String
    let other_name: String
    let full_name: String
    let date_of_birth: String
    let issue_date: String
    let expiry_date: String
    let nin: String
    let sex: String
    let card_number: String
    let age: Int
    let is_expired: Bool
    let fingerprint: FingerprintModel
    let warnings: [String]
}

struct FingerprintModel: Codable {
    let finger_index: Int?
    let minutiae_count: Int?
    let minutiae_bytes: Int?
    let sealed_block_bytes: Int?
}

class APIService {
    // SIMULATOR (backend on same Mac): http://127.0.0.1:8000
    // PHYSICAL DEVICE (same WiFi): http://YOUR_MAC_IP:8000
    static let baseURL = "http://127.0.0.1:8000"

    static func parseBarcode(payload: String, completion: @escaping (Result<ParseResponse, Error>) -> Void) {
        guard let url = URL(string: "\(baseURL)/parse") else { return }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let body = ParseRequest(payload: payload)
        request.httpBody = try? JSONEncoder().encode(body)

        URLSession.shared.dataTask(with: request) { data, response, error in
            if let error = error {
                completion(.failure(error))
                return
            }
            guard let data = data else {
                completion(.failure(NSError(domain: "No data", code: -1)))
                return
            }
            do {
                let result = try JSONDecoder().decode(ParseResponse.self, from: data)
                completion(.success(result))
            } catch {
                completion(.failure(error))
            }
        }.resume()
    }
}
